analyzer/evmtokenbalances/evm_token_balances.go

```diff
@@ -0,0 +1,304 @@
+	AccountAddrContextIdentifier string
+	AccountAddrContextVersion    int
+	AccountAddrData              []byte
+}
+
+func (m Main) getStaleTokenBalances(ctx context.Context, limit int) ([]*StaleTokenBalance, error) {
+	var staleTokenBalances []*StaleTokenBalance
+	rows, err := m.target.Query(ctx, m.qf.RuntimeEVMTokenBalanceAnalysisStaleQuery(), limit)
+	if err != nil {
+		return nil, fmt.Errorf("querying stale token balances: %w", err)
+	}
+	defer rows.Close()
+	for rows.Next() {
+		var staleTokenBalance StaleTokenBalance
+		if err = rows.Scan(
+			&staleTokenBalance.TokenAddr,
+			&staleTokenBalance.AccountAddr,
+			&staleTokenBalance.LastMutateRound,
+			&staleTokenBalance.Type,
+			&staleTokenBalance.Balance,
+			&staleTokenBalance.TokenAddrContextIdentifier,
+			&staleTokenBalance.TokenAddrContextVersion,
+			&staleTokenBalance.TokenAddrData,
+			&staleTokenBalance.AccountAddrContextIdentifier,
+			&staleTokenBalance.AccountAddrContextVersion,
+			&staleTokenBalance.AccountAddrData,
+		); err != nil {
+			return nil, fmt.Errorf("scanning stale token balance: %w", err)
+		}
+		staleTokenBalances = append(staleTokenBalances, &staleTokenBalance)
+	}
+	return staleTokenBalances, nil
+}
+
+func (m Main) processBatch(ctx context.Context) (int, error) {
+	staleTokenBalances, err := m.getStaleTokenBalances(ctx, MaxDownloadBatch)
+	if err != nil {
+		return 0, fmt.Errorf("getting discovered tokens: %w", err)
+	}
+	m.logger.Info("processing", "num_stale_token_balances", len(staleTokenBalances))
+	if len(staleTokenBalances) == 0 {
+		return 0, nil
+	}
+
+	ctxWithTimeout, cancel := context.WithTimeout(ctx, DownloadTimeout)
+	defer cancel()
+	group, groupCtx := errgroup.WithContext(ctxWithTimeout)
+
+	batch := &storage.QueryBatch{}
+
+	for _, stb := range staleTokenBalances {
+		// Redeclare `stb` for unclobbered use within goroutine.
+		staleTokenBalance := stb
+		group.Go(func() error {
+			m.logger.Info("downloading", "stale_token_balance", staleTokenBalance)
+			// todo: assert that token addr and account addr contexts are secp256k1
+			if staleTokenBalance.Type == nil {
+				m.logger.Warn("unrecognized token type", "token_addr", staleTokenBalance.TokenAddr)
+				batch.Queue(m.qf.RuntimeEVMTokenBalanceAnalysisUpdateQuery(),
+					staleTokenBalance.TokenAddr,
+					staleTokenBalance.AccountAddr,
+					staleTokenBalance.LastMutateRound,
+				)
+				return nil
+			}
+			balance, err := modules.EVMDownloadTokenBalance(
+				groupCtx,
+				m.logger,
+				m.cfg.Source,
+				staleTokenBalance.LastMutateRound,
+				staleTokenBalance.TokenAddrData,
+				staleTokenBalance.AccountAddrData,
+				*staleTokenBalance.Type,
+			)
+			if err != nil {
+				return fmt.Errorf("downloading token balance %s %s: %w", staleTokenBalance.TokenAddr, staleTokenBalance.AccountAddr, err)
+			}
+			if balance.Cmp(&staleTokenBalance.Balance.Int) != 0 {
+				var correction common.BigInt
+				correction.Sub(&balance.Int, &staleTokenBalance.Balance.Int)
```

Comment:
all this &asdf.Int :face_with_spiral_eyes:

Reply:
switching back to *big.Int for math. only using common.BigInt for db scanning for now

Reply:
Extemely optional and out of scope for this PR, but I'll say that `big.Int` has a butt-ugly (if presumably CPU-efficient) API. We could just add a few trivial methods to `BigInt` for basic arithmetics (`Plus()`, `Minus()`, `Times()`, `DivBy()`) and clean up our code.

Reply:
high five on big.Int math api dislike :raised_hands: